garage/tf/models/gaussian_mlp_model.py

```diff
@@ -0,0 +1,220 @@
+        :param std_parametrization: how the std should be parametrized. There
+          are a few options:
+            - exp: the logarithm of the std will be stored, and applied a
+                exponential transformation
+            - softplus: the std will be computed as log(1+exp(x))
+        :return:
+        """
+        Serializable.quick_init(self, locals())
+        super(GaussianMLPModel, self).__init__()
+
+        self.name = name
+        self._variable_scope = tf.variable_scope(
+            self.name, reuse=tf.AUTO_REUSE)
+        self._name_scope = tf.name_scope(self.name)
+
+        # Network parameters
+        self._input_dim = input_dim
+        self._output_dim = output_dim
+        self._hidden_sizes = hidden_sizes
+        self._learn_std = learn_std
+        self._init_std = init_std
+        self._adaptive_std = adaptive_std
+        self._std_share_network = std_share_network
+        self._std_hidden_sizes = std_hidden_sizes
+        self._min_std = min_std
+        self._max_std = max_std
+        self._std_hidden_nonlinearity = std_hidden_nonlinearity
+        self._hidden_nonlinearity = hidden_nonlinearity
+        self._output_nonlinearity = output_nonlinearity
+        self._std_parameterization = std_parameterization
+
+        # Tranform std arguments to parameterized space
+        self._init_std_param = None
+        self._min_std_param = None
+        self._max_std_param = None
+        if self._std_parameterization == 'exp':
+            self._init_std_param = np.log(init_std)
+            if min_std:
+                self._min_std_param = np.log(min_std)
+            if max_std:
+                self._max_std_param = np.log(max_std)
+        elif self._std_parameterization == 'softplus':
+            self._init_std_param = np.log(np.exp(init_std) - 1)
+            if min_std:
+                self._min_std_param = np.log(np.exp(min_std) - 1)
+            if max_std:
+                self._max_std_param = np.log(np.exp(max_std) - 1)
+        else:
+            raise NotImplementedError
+
+        inputs, outputs, model_info = self.build_model()
+        self._inputs = inputs["input_var"]
+
+        self._mean = outputs["mean"]
+        self._std = outputs["std"]
+        self._std_param = outputs["std_param"]
+        self._sample = outputs["sample"]
+        self._dist = model_info["dist"]
+
+        self._outputs = outputs
+
+    @overrides
+    def build_model(self):
+        """
+        Build the graph.
+
+        return:
+            inputs: a dict that contains input tensor
+            outputs: a dict that contains mean,
+                std, parameterized std and sample tensors
+            model_info: a dict that contains the distribution tensor
+        """
+        input_var = tf.placeholder(
+            shape=[None, self._input_dim],
+            dtype=tf.float32,
+        )
+
+        with self._variable_scope:
+            if self._std_share_network:
```

Comment:
Should we add the gaussian mlp model structure which has two outputs, one refers to mean, the other refers to std,  instead of one output that needs to be sliced to mean and std?

Reply:
They are not the same. A two headed-MLP is not the same as slicing the output of a single-headed MLP.

I think this should be split into several models. **Invariant: constructor arguments cannot change the architecture of the model, just parameters (e.g. hidden layers and sizes).**

Reply:
Oh, but I think two headed MLP is actually just a MLP connected with two MLPs right? I think chang was just saying that splitting the output layer?

Reply:
In my mind, two-headed means shared hidden layers, two output layers. You are variously proposing 
* Shared MLP, split the output
* One MLP (with hidden and output layers) outputs to two other MLPs (minimum 3 layers).